Feature/2020 05 30 terraform post (#604)
ordina-jworks/ordina-jworks.github.io: _posts/2020-05-30-terraform.md

1. Sync OpenShift service account credentials into Jenkins credential store
1. Roll-out Jenkins to add the new credentials by starting a Jenkins job executing an Ansible playbook
1. Create reverse proxy configuration in reverse proxy repository
1. Use a self-service portal to request access for Jenkins to update the configuration on the reverse proxy
1. Roll-out the reverse proxy configuration using a Jenkins job
1. Roll-out the environment configuration and application landscape to the new namespace through a Jenkins job that runs OC process and OC applies using the configuration.

This sums up the best-case scenario.
Since it requires multiple teams to perform disconnected tasks at different times, errors are frequent and slow to resolve.
Next to that, there is a lot of manual copy and paste involved to sync configuration from different sources into different repositories and systems.
Since any manual action, especially involving copying configuration between locations, is prone to errors, this introduces even more failure points.

> Combining the time it takes for tickets to be executed, copying the manual configuration to the correct locations and the debugging involved in getting the environment online results in lead times expressed in weeks.

Reviewer: Crazy, and not hard to believe with that many steps!

## Why use Terraform to supercharge infrastructure provisioning

As mentioned, one of the main drivers to look into automation for infrastructure was the move to the public cloud.
The reason why Azure was selected, might be discussed in a future blog post.

As Azure was selected, the logical choice for automating the infrastructure provisioning would be ARM templates.
An attempt was made at setting up the infrastructure with Terraform together with an ARM expert.
Unfortunately, multiple walls were hit due to the nature of the corporate setup on Azure.
One of the limitations that were hit, was the inability for ARM to make changes to resources that were provided by a central team.
More specifically, using a VNET that wasn't managed by ARM to deploy an Azure Kubernetes Service proved to be difficult or even impossible.

Using the CLI and scripts to create the setup was attempted as well, but this didn't fit the vision of a declarative setup for provisioning infrastructure.
Re-applying the same scripts either broke the setup or created additional, unneeded resources.
A clear no-go.

> As a third option, Terraform was investigated.

Reviewer: What is the difference here exactly compared to the first option? Seeing as Terraform was also investigated there.
I assume this option is about solely using Terraform?

Author: Added

It adhered to the vision of declarative definitions and all the components required by the landscape were available in the Azure Terraform provider.
Some basic setups were created and it showed great potential for the required setup.
So Terraform to the rescue!

## How to structure Terraform to allow for a two second tire change

One of the challenges of creating a Terraform setup is to determine a way to structure the code.
Looking online provided more questions than answers.
A lot of documentation can be found about how to structure a specific module or how to create modules, but almost no resources discuss how to structure the modules into logical, reusable components.
Until now!

The setup was divided into two different categories: managed services wrappers and standard provisioning components.

The first category is quite straight forward.
For every managed service that is being used, a module is created.
This module includes all the required Terraform resources for that service to operate.

> For example, the Key vault module contains the Azure Keyvault Terraform resource, but also the role assignments for the different Active Directory groups that require access to that Key vault.

Reviewer: I'd write it like this to distinguish between the generic term and the Azure keyed term.
```suggestion
For example, the key vault module contains the Azure Key Vault Terraform resource, but also the role assignments for the different Active Directory groups that require access to that key vault.
```

> Another example is the Container Registry module.

Reviewer: Same here as above.
```suggestion
Another example is the container registry module.
```

It contains the Azure Container Registry (ACR) Terraform resource, some role assignments and the Azure Monitoring Diagnostics Settings for making sure the ACR logs are shipped into the correct logs analytics bucket.
This abstraction allows for opinionated grouping of resources that are required for a managed service to operate.

The second category is a grouping of modules of the first category.
This set of modules are the ones that are actually provisioned during a deployment.

> Currently, only two of these modules exist a cluster module and a namespace module.

Reviewer: I'd add a colon here.
```suggestion
Currently, only two of these modules exist: a cluster module and a namespace module.
```

The cluster module contains all modules required to set up the shared resources for all environments with a similar purpose.
For example, all resources shared by all development environments.

> This module contains the Kubernetes cluster module, multiple resource group modules, a Keyvault module, some networking modules and some identity and role management modules.

Reviewer: ```suggestion
This module contains the Kubernetes cluster module, multiple resource group modules, a key vault module, some networking modules and some identity and role management modules.
```

This module is used once per cluster, meaning once for development, acceptance and production respectively.

The second module, namespace, is used to configure a namespace for a specific purpose: an environment for a development team, a specific testing environment or a rock-solid production environment.

> This module contains the DNS zone configuration, an API gateway module, another Keyvault module and a database module.

Reviewer: ```suggestion
This module contains the DNS zone configuration, an API gateway module, another key vault module and a database module.
```

These cluster and namespace modules are then used in a single Terraform module per cluster: so a single module for the development cluster together with all namespaces in that cluster.
This makes making changes to the infrastructure as easy as running a single Jenkins job executing that module.
Creating specific modules for development, acceptance and production also allows for the updates or upgrades of the configuration to be tested like application code before rolling it into production.

## How can you beat us

> With this setup, lead-time for provisioning an environment went down from 6 weeks of ticketing magic to just a 90 minute Jenkins run.

Reviewer: Very cool!

The decision to adopt Terraform still seems the correct one.
Infrastructure upgrades have become more and more stable over the adoption period and the different modules have matured to the point where they barely change anymore.

The work being done by the Azure Terraform provider community has helped tremendously.
They release new versions of the provider every week.
They aren't at feature parity with ARM, not by a long shot, but at the pace they are adding support for features, they will catch up very fast.

> The provider is also very stable and if an issue occurs (like [this](https://github.com/terraform-providers/terraform-provider-azurerm/issues/6525) AKS bug) it's fixed within the next release.

Reviewer: Open the link in a new tab to keep the reader with us. :)
```suggestion
The provider is also very stable and if an issue occurs (like [this](https://github.com/terraform-providers/terraform-provider-azurerm/issues/6525){:target="_blank" rel="noopener noreferrer"} AKS bug) it's fixed within the next release.
```

## Finally, two golden tips when using Terraform

The first is to start using remote state storage as soon as possible when using Terraform.
It provides an easy way to get an accurate Terraform plan which in turn provides an accurate overview of the actions Terraform will execute during the roll-out.

> The second one is roll-out the modules often and validate their effects by running tests against them.

Reviewer: ```suggestion
The second one is to roll out the modules often and validate their effects by running tests against them.
```
